### ISSUE TITLE
[Layer Upload] Allow to send SLD along with other files

### ISSUE THREAD

**Issue by afabiani** [geosolutions-it/unesco-ihp#109]
Currently GeoNode accepts SLDs also only if they are added to a ZIP file.

When trying to upload another file along with its SLD, the upload form prevents the user to proceed with an error message.

It would be good to be able to allow the upload form to recognize the sld in such cases also.

**Comment by meyerch** [comment]
Thank you for this improvement. 
Accepted and archived

**Comment by meyerch** [comment]
I am reopening this one as I tried to upload a few different rasters (so .tif + .sld together) and I sometimes get an error message:

![upload_layer_and_sld.PNG](https://waffleio-direct-uploads-production.s3.amazonaws.com/uploads/59c90ce9c2355f0011ed2e90/125516c66e82c728ace21e0d46fd84962c83879683e8a646b662f19a2d79347f2d46ef696097ec22ba416f43554d4ba5100c5b41f8afdf7aa3b439268a5c08edd56d5e71c03f03b0aac04ee155e57dd205d0f4ae805ec1ef3724e2898e4b391a011f8df95936b2d4.PNG)

I will send you by email the files for which upload resulted in this message.

**Comment by afabiani** [comment]
Hi @meyerch currently I updated fixes only to DEV instance.

Did not touch the PROD instance since the last update, a lot of time ago.

**Comment by meyerch** [comment]
Hi @afabiani 
Yes of course, I am talking about the Dev Instance only. 
I got this error message several times yesterday and this morning when trying to vector layer+.sld or raster+.sld on the Dev instance

**Comment by meyerch** [comment]
Hi @afabiani 
I no longer get the error message when uploading .sld along with other files of the layer. 

However, I ow have an issue with the final display: 
For the layer http://ihp-wins-dev.geo-solutions.it/layers/geonode:indicator_data_improved_water_sustai_cities1 , the legend is correctly displayed according to the .sld I uploaded along. 
However, nothing shows up on the map, and when scrollig done on the page, the "Styles" box indicates that there is no style associated to this layer.

![sld_layer_nostyle.PNG](https://waffleio-direct-uploads-production.s3.amazonaws.com/uploads/59c90ce9c2355f0011ed2e90/125516c66e82c728ace21e0d46fb989e1c8e82b08afb804dab4ee48d2543653f6b14b1247a80e03cfd183954560e1aa8445c0912adfc8b2ef0b63c738a580becd46b5d66da2f53fde6d640e258f075c95d8fa4bd8346c1ed3f23e48c89473a1d001f88ed10.PNG)

**Comment by meyerch** [comment]
After a few investigations, here are my observations for KML files

1/ when uploading .kml + .sld together: http://ihp-wins-dev.geo-solutions.it/layers/geonode:humanwaterstress_shp. 
It is like no .sld as been uploaded: the display is incorrect, the legend is incorrect, and the "Styles" box indicates a random style (in this case: 'a violet generic style'). 

2/ when uploading .kml + .sld together in a zipped file: http://ihp-wins-dev.geo-solutions.it/layers/geonode:humanwaterstress3. 
Display is correct, legend is correct, BUT the "Styles" box indicates 'no style'

3/  when upoading .kml, and then manually uploading the .sld (through Edit, Upload style): http://ihp-wins-dev.geo-solutions.it/layers/geonode:humanwaterstress. 
Display is correct, legend is correct BUT the "Styles" box indicates a random style (in this case: 'A red generic style')

**Comment by meyerch** [comment]
Do you prefer to open separate issues for each of my message on this thread?

**Comment by afabiani** [comment]
hi @meyerch well noted.
No need to create different issues for the moment. Let me examine the use cases you described in order to see if those side-effects are all related to the same cause.

**Comment by afabiani** [comment]
@meyerch after having a deeper look at the layers those are my findings:

1. _"For the layer http://ihp-wins-dev.geo-solutions.it/layers/geonode:indicator_data_improved_water_sustai_cities1 , the legend is correctly displayed according to the .sld I uploaded along."_

The SLD (at least the one associated to the layer) seems to be not compatible with the layer attributes. The server complains with the following message:
"The requested Style can not be used with this layer.  The style specifies an attribute of LATESVALUE and the layer is: geonode:indicator_data_improved_water_sustai_cities1"

2. _"when uploading .kml + .sld together: http://ihp-wins-dev.geo-solutions.it/layers/geonode:humanwaterstress_shp.
It is like no .sld as been uploaded: the display is incorrect, the legend is incorrect, and the "Styles" box indicates a random style (in this case: 'a violet generic style')."_

The error is the same as for _"when upoading .kml, and then manually uploading the .sld (through Edit, Upload style): http://ihp-wins-dev.geo-solutions.it/layers/geonode:humanwaterstress.
Display is correct, legend is correct BUT the "Styles" box indicates a random style (in this case: 'A red generic style')"_.

On GeoServer the style is correctly configured but somehow GeoNode was not able to correctly synchronize it. This is definitely a bug which I will fix ASAP.

3. _"when uploading .kml + .sld together in a zipped file: http://ihp-wins-dev.geo-solutions.it/layers/geonode:humanwaterstress3.
Display is correct, legend is correct, BUT the "Styles" box indicates 'no style'"_

Not sure if this was actually a KML or a Shapefile, however there's no error here. Just the uploaded SLD does not contain a "title" element.
This could be certainly improved by maybe displaying the name whenever the title is empty or absent.

**Comment by meyerch** [comment]
Hi @afabiani  
Sure, I am doing the tests and will let you know the outcomes. 

THanks

**Comment by meyerch** [comment]
Here are my findings on uploading sld @afabiani  (i did the test using a non admin account - najet.guefradj - and published all the uploads afterwards) : 

1/  For rasters: everything works well.  (http://ihp-wins-dev.geo-solutions.it/layers/geonode:Depth_to_Groundwater_Africa1_test35 / http://ihp-wins-dev.geo-solutions.it/layers/geonode:Depth_to_Groundwater_Africa1_test30 / http://ihp-wins-dev.geo-solutions.it/layers/geonode:Depth_to_Groundwater_Africa1_test3 )

2/ For kml, I have the same issue with all configurations: 
a/ uploading kml + sld at the same time (http://ihp-wins-dev.geo-solutions.it/layers/geonode:awali_test232), 
b/ uploading sld + kml in a zipped file (http://ihp-wins-dev.geo-solutions.it/layers/geonode:awali_test231),  
c/ uploading kml, then manually adding the sld (http://ihp-wins-dev.geo-solutions.it/layers/geonode:awali_test23)  
In all cases, I end up with the legend being correctly displayed, the correct style name in the "Styles" box, but no layer showing on the viewer. 

3/ For vector layers: everything works. (http://ihp-wins-dev.geo-solutions.it/layers/geonode:indicator_data_improved_water_sustai_cities1_3 / http://ihp-wins-dev.geo-solutions.it/layers/geonode:indicator_data_improved_water_sustai_cities1_2 / http://ihp-wins-dev.geo-solutions.it/layers/geonode:indicator_data_improved_water_sustai_cities1_1 )


2 bugs to report : 
1 For all the uploads done: when logged in as the person who uploaded the layers (najet.guefradj), the brush icon to edit the Style  above the viewer is systematically greyed and deactivated... As an admin, the brush is activated on all layers so I can modify them. Contributors should be able to edit the layers they have uploaded. 

2 I (admin) have approved and published all uploaded layers. It is confirmed by the fact that no layer has the mention "to be approved" or "not published" in the layers' list. However, as an anonymous user, I don't see any of the layers uploaded..

**Comment by afabiani** [comment]
@meyerch 

About the "awali_test232" layer you uploaded I taken a look at the GeoServer logs, which I'm gonig to report here below:

<?xml version="1.0" encoding="UTF-8" standalone="no"?><!DOCTYPE ServiceExceptionReport SYSTEM "http://ihp-wins-dev.geo-solutions.it/geoserver/schemas/wms/1.1.1/WMS_exception_1_1_1.dtd?access_token=GPUPr02UBE3Hd31FJ7jqwB2EUxNa3Q"> <ServiceExceptionReport version="1.1.1" >   <ServiceException>
      The requested Style can not be used with this layer.  The style specifies an attribute of Name and the layer is: geonode:awali_test232
</ServiceException></ServiceExceptionReport>

Basically it looks like the SLD you are trying to upload along with the layer is not compatible with the layer itself. How did you produced the style? Is this working with other tools?

About the 2 other bugs you mentioned I'll provide you a reply ASAP.

**Comment by afabiani** [comment]
@meyerch those are my findings about the other 2 bugs you mentioned:

1. Brush deactivated; this was due to an old security settings which was not updated to the new GeoServer 2.13.x - It  should be good now.

2. This issue seems to be related again to a misconfiguration of the security groups. Maybe we didn't clean up the Groups in DEV server. It looks like the SDG6 Group Profile you selected for the new layers is a PRIVATE group, and currently we have setup GeoNode to hide resources belonging to private groups to everyone except the owners and the group members. Is this workflow still valid?

**Comment by meyerch** [comment]
@afabiani 
For the "awali_test232" layer, it is interesting because you mentionned the same problem in previous tests I did for .kml. 
In all cases, the sld style for the kml was creating using QGIS, so there might be an issue from here. However, kml are correctly displayed in QGIS.
Let me know if you need anything from me

**Comment by afabiani** [comment]
@meyerch Uhm, ok, let me double check the SLD on GeoServer. Maybe it is an issue with some of the layer attributes names? I will provide you feedbacks as soon as possible.

**Comment by afabiani** [comment]
@meyerch I already managed to spot the issue, but don't know exactly how this happened, probably a problem on QGIS side during the conversion of SLDs.

Let me explain briefly...

First of all I uploaded the same fixed SLD to this layer

http://ihp-wins-dev.geo-solutions.it/layers/geonode:awali_test231

and, as you can see, it is working as expected now.

Basically the issue was in the Rules Filter. All the rules are trying to filter against a Property (aka a Layer Attribute) called "Name" with the first cap letter, while the Layer actually has all the attributes lowercase (or at least this is how GeoServer ingested it).

So basically the SLD starts working well by just changing "Name" to "name".

**Comment by meyerch** [comment]
For the 2nd message: 
1/ For the brush, it works now, thanks.

2/ OK, sorry for this mistake. Is it possible to setup the 'Private' groups' in the same way as the  'Public on invitation' groups? I.e. resources belonging to private groups can be seen by the owners and groups members as long as they are not published, and can be seen by everyone else when published? 
So basically, 'Public on invite' and 'Private' groups would be the same, with only a difference in name.

**Comment by afabiani** [comment]
@meyerch about 2/ that means we need to tweak several parts of the GeoNode core logic, which leads us to diverge from current GeoNode implementations.

I need final confirmation that you still need to apply this change.

**Comment by meyerch** [comment]
hi again @afabiani 
ok, then we do not change anything and we keep the workflow for "private" groups as it is now. 
Thank you
